### ISSUE TITLE
[DX] [BC] Deprecate, reduce use of, or remove DBTNG

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#61]
We should considering deprecating/reducing use of/removing DBTNG in Backdrop to accomplish the following goals:
- Improve performance by removing abstraction around constructing queries.
- Improve learnability by avoiding a Drupal-specific abstraction layer and syntax.

This introduces a need for substitutes in a number of places however:
- Alterability: Remove the ability to alter queries. No more alterableQueries or db_rewrite_sql(). Instead, any query that needs to be altered instead utilizes a View, that can be altered to affect the query. Or, avoid altering entirely through...
- Access control: Make access control queries explicitly defined when creating Views. Instead of automatically restricting listings, require listings to have access control explicitly defined (similar to explicitly defining the "published" state).
- Database connections: We can still use PDO to connect to the database. It's just as fast as a mysqli connection in benchmarks, and makes it so we can use a consistent syntax for doing queries, regardless of the back end.

Things we would NOT remove or build out later include:
- Schema API: As long as we support PostgreSQL (which we should for the initial release at least), Schema API is still needed to define and update tables.
- Entity abstraction: We should focus on ensuring entities are loaded and modified through the entity API alone, making it so it's still possible to cache and store entities in a swappable mechanism. Although things like MongoDB are probably not the target audience, efficient per-entity caching definitely is, so this benefits us for the time being.

The initial effort should be to provide a database system that can update usage of db_query() throughout core with a lightweight replacement directly against PDO. We can work on removing db_select/update/insert/delete queries over time and then remove them (and DBTNG) when finished.

**Comment by jenlampton** [comment]
> Haven't met any D7 developers who dislike dbtng

Really? I have. I've also met a ton of developers of other systems who know MySQL, and get really frustrated by Drupal's DBTNG. 

I myself often spend hours trying to read documentation on how to get DBTNG to do something complex, then I give up, and write my own MySQL query in minutes.

**Comment by fgm** [comment]
In that perspective, would you go to plain PHP mysqlnd API ? PDO ? DBAL ? something else ?

Given the history of the PHP projects with low-level APIs, the mysqlnd option seems a bit risky. DBAL would come with a nicer than current API, especially regarding schema migrations.

**Comment by jlfranklin** [comment]
I appreciate DBTNG for what it's trying to accomplish.  The three major benefits I see are:

- Translation of intent to SQL that works for the underlying layer.  For example, SQLite doesn't have `TRUNCATE`.  You have to use `DELETE FROM` without a `WHERE` clause.  It doesn't handle date conversions, like `FROM_UNIXTIME()`, but it could. (#2890)
- Helps prevent lazy devs from writing dangerous queries such as `db_query("UPDATE {foo} SET bar=$sql_injection_vulnerable_variable")`.
- Makes adding conditional parts of a query easier to code, like this:

```php
    if (empty($foo)) {
      $query->condition('bar', $baz);
    }
```

I agree with @jenlampton, complex queries are hard to do, and can be really hard to read.  I often put the intended SQL in a comment before the block of code so others coming after me (including future me) have a fighting chance of understanding what it's trying to do.  

For all its warts, I don't know that I could do a better job designing a database abstraction layer, and I have tried.  It would be easier to convert Backdrop's database layer to an ORM (think: RoR's ActiveRecord) instead of a safe SQL construction kit, but that would have huge ripple effects across all of core and contrib.

I'm not in favor of just dumping it and returning to `db_query()` all the time.  Without something similar, the number of SQL-injection security updates will spike.  (Oh, and it's a critical part of Silkscreen.)

If someone has some ideas on how to improve DBTNG, I'm all ears.

**Comment by fgm** [comment]
Beyond its actual internal design, the main issue I have with DBTNG as it stands is that it could easily be a normal PHP package usable via Composer, but a few implementation details prevent it from being one. 
Then again, Composer deployments are possibly more of a thing for Drupal than Backdrop, considering the respective targets.

**Comment by klonos** [comment]
Just wanted to chime in and say that as a novice dev, I personally find DBTNG easier to read/understand and then also tinker with, rather than using "direct" MySQL queries via `db_query()`. In my head, "direct" MySQL queries add another thing that Drupal/Backdrop newbies will need to learn (SQL) in order to contribute, so basically adds to the learning curve.

My 2c.

**Comment by jenlampton** [comment]
> I often put the intended SQL in a comment before the block of code so
others coming after me (including future me) have a fighting chance of
understanding what it's trying to do.

I always do this in my custom modules! :)

The problem with DBTNG for new people is they they can see and manipulate
simple queries, but they can see and manipulate simple queries in MySQL,
too. Complex queries is where they get stuck because of lack of (or
inadequate) documentation -- and that trips up the pros too. At least with
MySQL we have all the same resporses for help as the rest of the world, so
Google and stack overflow can come to our rescue. With DBTNG we're stuck
relying on the 1 or 2 people who really understand it to write good docs.

I'm also skeptical that DBTNG is any more secure than db-query() for
preventing SQL injection. The only real protection it offers is that it
looks unfamiliar to people, so they will be more likely to read the docs,
or copy an existing implementation. This makes the difficulty into a
feature, and that just feels like we're doing it wrong. There were some
huge security holes in DBTNG queries in core -- and there probably still
are.  If the experts can't figure out how to use it correctly, I have
little hope that everyone else will.

I'm curious to see what replacing the DBTNG query builder with the Views
query builder looks like. That's the direction I'd like to go for
Deprecating DBTNG, and I'm hoping that will make it possible to provide
backwards compatibility for a full major release cycle.

**Comment by klonos** [comment]
> ... Complex queries is where they get stuck because of lack of (or inadequate) documentation -- and that trips up the pros too.

That's true.

>At least with MySQL we have all the same resporses for help as the rest of the world, so Google and stack overflow can come to our rescue.

...also true.

**Comment by klonos** [comment]
If there is consensus about deprecating DBTNG (with whatever/something), shouldn't we be discouraging its usage by adding watchdog warnings in functions like `db_select()` / `db_update()` / `db_delete()` etc.? I mean, the less people are using them, the less work will need to be done when we actually deprecate them.

**Comment by jenlampton** [comment]
Maybe? I think we shouldn't start with the notices until we're sure a out
this - which means we need to prove bc is possible. As it is now the
removal wohldnt happen till at least 2.x, more likely 3.x, so we have
plenty of time.

**Comment by twmfc** [comment]
I'm a bit player but in D7 I use code like:

$result = db_select('node', 'n')->fields('n', array('nid, title'))->$range(0, 10)->addTag('node_access');

a fair bit and I'd be disappointed to lose this in BD

**Comment by herbdool** [comment]
> shouldn't we be discouraging its usage by adding watchdog warnings in functions like db_select() / db_update() / db_delete() etc.?

I don't think there was consensus yet. Even then this would result in spamming the watchdog since the queries are used all over core right now. First, they should be eliminated from core. Second, it seems this needs to be done first as well:

> The initial effort should be to provide a database system that can update usage of db_query() throughout core with a lightweight replacement directly against PDO.

**Comment by jenlampton** [comment]
Again: we currently have three query builders in core. Replacing DBTNG with
something else won't solve that. We need to switch to using Views for
anything that needs access control, and see what other use-cases we have
that need to be addressed.

We don't need to replace all the features that DBTNG brought us, we only
need to solve the use-cases that exist for Backdrop. Database abstraction,
for example, is no longer necessary.

**Comment by herbdool** [comment]
I'm personally not invested in removing or keeping DBTNG. Though I'm willing to put money on it not being removed any time soon. Sounds like there would be a lot of work and I don't imagine it's a high priority for anyone developing for Backdrop.

I read this thread with interest, including the data on how much impact it has on query time. But I didn't see any data on how it impacts page load time from the end user's perspective. Can we put a number to it roughly? Is 10%, 50%? I'm betting it's at the low end - how does that impact the priority of this? Are there other things we can do in core that are easier and have an equal or bigger impact on performance?

**Comment by jenlampton** [comment]
My interest in this issue is primarily for learnability and simplicity and
consistency, not speed, since DBTNG was one of the hardest things for new
Drupal 7 adopters to learn. For me, the performance gain was a nice side
affect that also lines up with our principles.

However, I think until 2020-2021 our focus should be on improvement in
compatibility with Drupal 7, and this issue should be postponed until at
least then.

When we do get to this, we should aim for backwards compatibility for at
least one release cycle, and if we can get there, it may be posible to
shift DBTNG to contrib!

**Comment by klonos** [comment]
> However, I think until 2020-2021 our focus should be on improvement in compatibility with Drupal 7, and this issue should be postponed until at least then.

👍 ...our focus till end of 2021 should be to smoothen the migration/upgrade process of D6/D7 sites to Backdrop.

**Comment by twmfc** [comment]
_We need to switch to using Views for anything that needs access control, and see what other use-cases we have that need to be addressed._

Perhaps I'm missing the point here, but my understanding is that Views are just a further abstraction away from the raw sql used to access the database (MySQL only in BD but a range of DB's in Drupal and Silkscreen).

Using views will traverse a path that looks like this:

Views--->DBTNG--->sql--->db

Specifically, as I understand it, Views will utilise hook_query_alter() to add tags, via DBTNG, to build up the final sql query. 

Therefore, I do not see Views as a replacement for DBNTG; rather I see Views as a query builder that makes use of DBTNG along with providing a GUI.

Pleased to see that the consensus appears to be to defer this until further down the line as I'm sure this will better reward all of the hard work the Team have done with BD thus far!

**Comment by jlfranklin** [comment]
> Using views will traverse a path that looks like this:
>
> Views--->DBTNG--->sql--->db

That is accurate.  There are a couple things that Views does with dates that could be pushed down into the database layer.  (#2890)  Beyond that, Views uses DBTNG to build the queries.

**Comment by jenlampton** [comment]
> Views are just a further abstraction away from the raw sql used to access the database

yes, exactly!

Views does not need to use DBTNG in order to achieve both access control and alterable queries, as it existed before DBTNG did. IF we can switch everyone over to using Views instead of DBTNG directly, we would have abstracted away the DBTNG, making it easier to remove that part.

That would make our types of queries look like:
1) db_query--->sql--->db
2) Views--->sql--->db

...rather than:
1) db_qurey--->sql--->db
1) DBTNG--->sql--->db
1) Views--->DBTNG--->sql--->db

The views query builder would certainly need changes - but I suspect we may be able to do **that** in a way that would have the least impact on contrib. Since views exist now, maintainers could begin switching things over to views at any time. If that were our path away from DBTNG, then both could safely exist for a long while, and changes to Views may not need any further changes to contrib.

Caveat: This of course, is all theoretical until we try to write some code :)
